### PR TITLE
a custom datetime shard algorithm used to shard tables by days/weeks/months/QUARTER_OF_YEAR

### PR DESCRIPTION
**Pull request by SanmuWangZJU**
Ref #5282.

Changes proposed in this pull request:
- supply a sharding algorithm used to shard tables by days/weeks/months/QUARTER_OF_YEAR/...
- usage examples
  - shard by QUARTER_OF_YEAR (按季度分表)
  ``` yaml 
      actualDataNodes: ds.t_order_20${10..20}0${1..4}
      tableStrategy:
        standard:
          shardingColumn: created_at
          shardingAlgorithm:
            type: CUSTOM_DATE_TIME
            props:
              datetime.format: yyyy-MM-dd HH:mm:ss
              table.suffix.format: yyyyQQ
              datetime.lower: "2010-01-01 00:00:00.000"
              datetime.upper: "2020-12-31 00:00:00.000"
              datetime.step.unit: Months
              datetime.step.amount: 3
  ```
- config detail can refer to java doc of `org.apache.shardingsphere.core.strategy.algorithm.sharding.CustomDateTimeShardingAlgorithmTest`

### CODE REVIEW

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ *
+ * <p>detail explain for each char in datetime.format and table.suffix.format can refer {@link TemporalField}.
+ *
+ * <p>datetime.lower and datetime.upper: if app query with only half bound, lower and upper helps to build other half bound,
+ * datetime.lower must be specified and datetime.upper has a default value to {@link LocalDateTime#now}
+ * (default value of datetime.upper could only be used when query sql needn't get result that time larger than query time).
+ *
+ * <p>datetime.step.unit and datetime.step.amount used for calculate tables for range shard, datetime.step.unit is name of
+ * {@link ChronoUnit}, default unit is Days and amount is 1, amount + unit should not be larger than but close to your shard range.
+ *
```
Comment (tristaZero) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:60]: I like this detailed java doc.

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        CustomDateTimeShardingAlgorithm shardingAlgorithm = new CustomDateTimeShardingAlgorithm();
+        shardingAlgorithm.getProperties().setProperty("datetime.format", "yyyy-MM-dd HH:mm:ss");
+        shardingAlgorithm.getProperties().setProperty("table.suffix.format", "yyyyQQ");
+        shardingAlgorithm.getProperties().setProperty("datetime.lower", "2016-01-01 00:00:00.000");
+        shardingAlgorithm.getProperties().setProperty("datetime.upper", "2021-12-31 00:00:00.000");
+        shardingAlgorithm.getProperties().setProperty("datetime.step.unit", "Months");
+        shardingAlgorithm.getProperties().setProperty("datetime.step.amount", "3");
+        StandardShardingStrategyConfiguration shardingStrategyConfig = new StandardShardingStrategyConfiguration("create_time", shardingAlgorithm);
+        this.shardingStrategy = new StandardShardingStrategy(shardingStrategyConfig);
+
```
Comment (tristaZero) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java:63]: A redundant blank line is recommended to remove.

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    @Test
+    public void assertUpperHalfRangeDoSharding() {
+        Range<String> rangeValue = Range.atMost("2019-09-01 00:00:00");
+        List<RouteValue> shardingValues = Lists.newArrayList(new RangeRouteValue<>("create_time", "t_order", rangeValue));
+        Collection<String> actual = shardingStrategy.doSharding(availableTables, shardingValues, new ConfigurationProperties(new Properties()));
+        assertThat(actual.size(), is(15));
+    }
+
+    @Test
+    public void testFormat() {
```
Comment (tristaZero) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java:106]: Considering the consistent function name, `assertFormat` is better.

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ */
+public class CustomDateTimeShardingAlgorithmTest {
+
+    private final List<String> availableTables = new ArrayList<>();
+
+    private StandardShardingStrategy shardingStrategy;
+
+    @Before
+    public void setup() {
+        CustomDateTimeShardingAlgorithm shardingAlgorithm = new CustomDateTimeShardingAlgorithm();
```
Comment (tristaZero) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java:54]: This is a great practice t shard by quarter. Besides, could you provide another test case `like sharding by month or day` to present its flexibility to our users? 
You know, a cool algorithm needs more examples to express its features.

**Comment by tristaZero** [comment]
Hi @SanmuWangZJU 
What you did is greater than expected! 
There are just a few suggestions in the comment section, please review them and look forward to your reply. :-)

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java (hunk 1 above)
Comment (kimmking) [reply sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:60]: Me too

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            }
+        }
+    }
+
+    private void verifyProperties() {
+        Preconditions.checkNotNull(properties.getProperty(DATE_TIME_FORMAT));
+        Preconditions.checkNotNull(properties.getProperty(TABLE_SUFFIX_FORMAT));
+        Preconditions.checkNotNull(properties.getProperty(DEFAULT_LOWER));
+        stepUnit = properties.getProperty(STEP_UNIT) == null
+                ? stepUnit = ChronoUnit.DAYS
```
Comment (kimmking) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:153]: remove tepUnit =

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    private String formatForDateTime(final LocalDateTime localDateTime) {
+        return localDateTime.format(DateTimeFormatter.ofPattern(properties.get(TABLE_SUFFIX_FORMAT).toString()));
+    }
+
+    private void mergeTableIfMatch(final LocalDateTime dateTime, final Collection<String> tables, final Collection<String> availableTargetNames) {
+        String suffix = formatForDateTime(dateTime);
+        availableTargetNames.parallelStream().filter(tableName -> tableName.endsWith(suffix)).findAny().map(tables::add);
+    }
+
+    private void checkInit() {
```
Comment (kimmking) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:137]: why add sync block here？

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+ *
+ * <p>properties defined here:
+ *
+ * <p>datetime.format: the datetime format used by applications, must can be transformed to {@link LocalDateTime},
+ * used by {@link LocalDateTime#parse(CharSequence, DateTimeFormatter)}.
+ *
+ * <p>table.suffix.format: suffix for sharded tables, used by {@link LocalDateTime#format(DateTimeFormatter)},
+ * examples:
+ * suffix=yyyyQQ means shard by {@link IsoFields#QUARTER_OF_YEAR};
+ * suffix=yyyyMMdd means shard by {@link ChronoField#DAY_OF_YEAR}.
```
Comment (kimmking) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:50]: yyyyMM is also a common style.

**Comment by kimmking** [comment]
Good job.
And one suggestion here:
1. support yyyyMM for users, it may be the most popular style for this case.

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java (hunk 7 above)
Comment (SanmuWangZJU) [reply sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:50]: yes, it support yyyyMM, I will add this example to java doc and provide corresponding test cases

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java (hunk 6 above)
Comment (SanmuWangZJU) [reply sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithm.java:137]: sync block here is used to ensure `verifyProperties()` method exactly execute once when multi-thread call `doSharding()`, inspired by Double-Checked Locking with Singleton

**Comment by SanmuWangZJU** [comment]
Hi @tristaZero  @kimmking 
Thanks for your suggestions, I have add  test case for yyyyMM(shard by Month) and fix some format issues.
Looking forward to your further suggestions :)

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java (hunk 4 above)
Comment (SanmuWangZJU) [reply sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/algorithm/sharding/CustomDateTimeShardingAlgorithmTest.java]: I have just add test case for sharding by month, looking forward to further suggestions

**Comment by tristaZero** [comment]
Hi @SanmuWangZJU 
I am glad you contribute this practical idea to the os community! Hope more people can benefit from it.

**Comment by tristaZero** [comment]
@SanmuWangZJU 
Your PR told me you are veteran, welcome to play a valuable role in ShardingSphere community. 😉

**Comment by SanmuWangZJU** [comment]
> @SanmuWangZJU
> Your PR told me you are veteran, welcome to play a valuable role in ShardingSphere community. 😉

Thanks, hope more of my ideas benefit the community.

**Pull request merged**